File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        int upperbound = INT_MAX;
        int lowerbound = INT_MIN;
        return dfs(root, upperbound, lowerbound);
    }
    bool dfs(TreeNode* root, int upperbound, int lowerbound){
        if(!root){
            return true;
        }
        else if(root->val >= upperbound || root->val <= lowerbound){
            return false;
        }
        else{
            return dfs(root->left, root->val, lowerbound) && dfs(root->right, upperbound, root->val);
        }
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        q = []
        res =[]
        if root:
            q.append(root)
        while q: 
            size = len(q)
            for i in range(len(q)):
                node = q[0]
                if i == size - 1:
                    res.append(node.val)
                q.pop(0)
                if node.left: 
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return res;  
        